push/push-reduce-footprint-and-messaging.md

```diff
@@ -0,0 +1,92 @@
+# Abstract
+
+This document outlines a modularization approach of the UPS, for better scalablitly and lower footprint when running large push workloads
+
+## Goals
+
+* better scalablitly through modularization of the network heavy sender processes
+* lower footprint, by breaking workloads out of the _core_ Push Server 
+
+## Terms
+
+ - Kafka: [distributed and partioned commit log, with message system semantics](https://speakerdeck.com/matzew/eclipse-microprofile-and-apache-kafka?slide=9)
+ - Pod: Group of containers deployed on the same host
+ - JavaEE: framework for writing managed servers with build-in features for persistence, transaction and dependency injection. Applications are typically deployed in an archive (`WAR`).
+ - WildFly Swarm: Java EE microservice tool
+ - nginx: https server and proxy 
+ - System Resources: CPU or memory
+
+## Problem
+
+The UPS 1.2.x line is hard to scale and requires a good amount of system resources. Currently all of the push functionality is baked into one bigger `WAR` file:
+ - housekeeping/crud for all the metadata for PushApplications, their Variants and their devices.
+ - Lookup of the device tokens for a push deliver
+ - Push delelivery to the different 3rd party push networks (such as APNs or FCM)
+ - simple analytics about Push Notifications
+
+In the past we have seen problems when a larger amount of push notifications were sent. Problems with lot's of push delivery connections to the different 3rd networks, e.g. hanging/stale connections, did cause memory issues, which eventually did cause the entire server (or system) to fail.
+
+Another problem is that the current architecture lacks of good ways to (horizontally) scale **only** the busy parts of the _push system_. In a perfect world it would be trivial to scale up only those parts that are currently busy. However, currently it's not possible, only the entire server can be replicated.
+
+## Proposed solution
+
+Refactoring the current UPS into a more modular system, or a system of systems.
+
+### Apache Kafka
```

Comment:
how much resource will Kafka require?

Also given that there is a messaging service from middleware is being developer for OpenShift (is it called EnMasse?), should we consider this part replaceable?

Reply:
Kafka cluster means a few nodes for Kakfa, and Zookeeper. Each node needs a few hundred MiB too.  

I'd think that Kafka is a "Mobile System level" installation - since it does offer much value for all other services as well: it can be the centralized data hub.

yes, I use the EnMasse Kafka templates for Openshift